In [1]:
# importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
#importing dataset
df = pd.read_csv('fuel_ferc1.csv')
df.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


In [4]:
df.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [5]:
df.describe()

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count      29523.000000  29523.000000     2.952300e+04         29523.000000   
mean         118.601836   2005.806050     2.622119e+06             8.492111   
std           74.178353      7.025483     9.118004e+06            10.600220   
min            1.000000   1994.000000     1.000000e+00             0.000001   
25%           55.000000   2000.000000     1.381700e+04             1.024000   
50%          122.000000   2006.000000     2.533220e+05             5.762694   
75%          176.000000   2012.000000     1.424034e+06            17.006000   
max          514.000000   2018.000000     5.558942e+08           341.260000   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count               29523.000000                  2.952300e+04   
mean                  208.649031                  9.175704e+02   
std                  2854.490090                  6.877593e+04   
min                  -276.080000                 -8.749370e+02   
25%                     5.207000                  3.778500e+00   
50%                    26.000000                  1.737100e+01   
75%                    47.113000                  4.213700e+01   
max                139358.000000                  7.964521e+06   

       fuel_cost_per_mmbtu  
count         29523.000000  
mean             19.304354  
std            2091.540939  
min             -41.501000  
25%               1.940000  
50%               4.127000  
75%               7.745000  
max          359278.000000

## Question 1

If you’re given two lists:

A = [1,2,3,4,5,6], B = [13, 21, 34]

The task is to create a list with the elements of A and B in a single dimension with output:

A_B = [1,2,3,4,5,6,13, 21, 34]

Which of the following option is the best way to create this list?

In [6]:
A = [1,2,3,4,5,6]
B = [13,21,34]
A.extend(B)
A

[1, 2, 3, 4, 5, 6, 13, 21, 34]

## Question 2
How do you create an identity matrix in python?

In [7]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

## Question 3

Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [8]:
fuel_type = df.groupby("fuel_type_code_pudl")

fuel_type["fuel_cost_per_unit_burned"].describe()

count         mean           std     min      25%  \
fuel_type_code_pudl                                                        
coal                  8547.0    67.421830    943.771433    0.00  23.9585   
gas                  11486.0    13.659397    163.286631    0.00   3.2300   
nuclear                818.0  4955.157002  14291.929749    0.00   1.6015   
oil                   8064.0   168.877086   2406.101479    0.00  37.6100   
other                  167.0    18.253856     29.016882 -119.04   2.6325   
waste                  441.0    19.518122     29.988017 -276.08  10.3350   

                         50%       75%         max  
fuel_type_code_pudl                                 
coal                 33.3180  44.66800   44033.320  
gas                   4.6500   6.96000   10490.000  
nuclear              32.5200  50.85525   69785.215  
oil                  68.9455  98.68925  139358.000  
other                 9.0640  20.05500     134.547  
waste                21.9330  32.65000     132.291

In [9]:
fuel_type["fuel_cost_per_unit_burned"].describe()["mean"]

fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: mean, dtype: float64

#### Gas has the lowest average fuel_cost_per_unit_burned with a value of 13.659397

## Question 4
What is the standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [10]:
Energy_per_unit = df["fuel_mmbtu_per_unit"].describe()
Energy_per_unit

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

#### The standard deviation is 10.60 and the 75 percentile is 17.01

## Question 5
What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [11]:
pd.DataFrame.skew(df["fuel_qty_burned"])

15.851495469109503

In [12]:
pd.DataFrame.kurtosis(df["fuel_qty_burned"])

651.3694501337732

#### The skewness is 15.85
#### The Kurtosis is 651.37

## Question 6

Which feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places?

(Enter answer in the format: Feature: xxx, Total: xxx, Percent: xxx)

In [13]:
df.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [14]:
percentage = (180/29523) * 100
percentage

0.609694136774718

In [15]:
round(percentage,2)

0.61

#### Feature: fuel_unit, Total: 180, Percent: 0.610

## Question 7

The feature with missing values falls under what category? What missing value imputation technique would you use?

In [16]:
df['fuel_unit'].dtype

dtype('O')

In [17]:
df['fuel_unit'].mode()

0    mcf
dtype: object

#### Categorical and mode imputation

## Question 8
Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [18]:
correlation = df.corr()

correlation

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

In [19]:
correlation['fuel_cost_per_unit_burned']

utility_id_ferc1               -0.037863
report_year                     0.013599
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_unit_burned       1.000000
fuel_cost_per_unit_delivered    0.011007
fuel_cost_per_mmbtu            -0.000437
Name: fuel_cost_per_unit_burned, dtype: float64

In [20]:
correlation['fuel_cost_per_unit_burned'].sort_values(ascending=True)

utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

#### ''fuel_qty_burned'' and ''fuel_mmbtu_per_unit" are the features that has the second and third lowest correlation with the Fuel Cost Per Unit Burned.

## Question 9
For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [21]:
df_info = df.groupby(["report_year", "fuel_type_code_pudl"])
info = df_info["fuel_cost_per_unit_burned"].sum()
info.head(40)

report_year  fuel_type_code_pudl
1994         coal                    14984.572
             gas                     10792.496
             nuclear                227983.354
             oil                      9362.194
             other                     116.737
             waste                      52.762
1995         coal                    14571.785
             gas                      9087.905
             nuclear                384152.639
             oil                     17858.281
             other                     183.577
             waste                    -192.699
1996         coal                    12694.803
             gas                      1467.390
             nuclear                 66657.887
             oil                      5357.028
             other                     160.075
             waste                      88.092
1997         coal                    43742.178
             gas                      1518.471
             nuclear       

#### for 1994, coal cost per unit burned = 14984.572
#### for 1998, coal cost per unit burned = 11902.597

In [22]:
Percentage_change = ((11902.597 - 14984.572) / 14984.572) * 100
Percentage_change
round(Percentage_change)

-21

#### The percentage change in the fuel cost per unit burned in 1998 compared to 1994 is "-21"

## Question 10
Which year has the highest average fuel cost per unit delivered?

In [23]:
rep_yr = df.groupby("report_year")
rep_yr["fuel_cost_per_unit_delivered"].describe()

count          mean            std      min      25%      50%  \
report_year                                                                   
1994         1235.0     63.636060    1288.748659  -58.800  2.32600  13.1560   
1995         1201.0     32.735269     173.849821  -85.640  1.80600   5.1140   
1996         1088.0   9196.705948  244767.437600  -44.050  2.93500  16.2700   
1997         1094.0  11140.197239  256945.853368  -38.972  2.94650  14.4515   
1998         1107.0    287.154420    8172.214382  -43.747  2.40000   3.4750   
1999         1050.0     25.551627     137.826776  -32.970  2.52700   3.5750   
2000         1373.0    985.362877   31576.526616 -174.670  4.37000  19.4700   
2001         1356.0     60.050396     947.767770  -37.790  4.35950  21.2150   
2002         1205.0     47.594361     906.266622  -60.290  3.72400  17.8940   
2003         1211.0     55.663493     996.166868 -118.340  5.93000  21.9670   
2004         1192.0    139.524275    3848.790040 -313.410  6.51500  20.9000   
2005         1269.0     41.438184     247.442252  -31.501  8.68100  23.8610   
2006         1243.0     38.657484      68.938342  -24.295  7.58500  21.7870   
2007         1264.0     43.325023     148.371871 -488.054  7.48925  22.9085   
2008         1228.0     58.588197     314.507364 -874.937  9.28175  23.5950   
2009         1222.0    652.694163   20888.844544    0.000  5.38550  26.6120   
2010         1261.0     91.862105    1282.253549   -7.295  5.41400  24.9000   
2011         1240.0     59.774667     296.896180  -43.419  5.18075  24.5155   
2012         1243.0     60.994502     313.148171  -91.900  3.56550  25.4000   
2013         1199.0    172.307591    3917.510920  -93.470  4.61050  24.1770   
2014         1171.0    192.737183    4019.515097    0.000  5.50200  29.3400   
2015         1093.0    326.535511    5571.037893 -170.187  3.38000  12.0690   
2016         1034.0    103.901761    2163.468884  -40.725  3.08925   6.3795   
2017          993.0     46.196861     209.871096    0.000  3.35100   6.1190   
2018          951.0    499.269966    8976.783325    0.000  3.33000   5.8400   

                   75%          max  
report_year                          
1994          30.91000    45115.000  
1995          29.79000     2396.770  
1996          30.57250  7793482.000  
1997          29.10400  7964521.000  
1998          24.78000   270604.000  
1999          26.12300     2891.100  
2000          36.28800  1168000.000  
2001          35.78250    34381.000  
2002          31.96000    31416.000  
2003          37.44100    34518.000  
2004          45.84025   132884.000  
2005          62.13000     8651.000  
2006          68.49700     1431.000  
2007          72.66500     4850.872  
2008          80.97925    10490.000  
2009          71.45950   730166.000  
2010          86.36000    44033.320  
2011          92.69675     9545.000  
2012          99.53500    10129.000  
2013          90.49000   135315.000  
2014         103.24750   136410.000  
2015          67.46000   140706.000  
2016          54.04875    69440.000  
2017          62.93000     3584.335  
2018          70.43900   245652.000

In [24]:
rep_yr["fuel_cost_per_unit_delivered"].describe()['mean'].sort_values(ascending=False)

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: mean, dtype: float64

#### The year that has the highest average fuel cost per unit delivered is 1997